Harden nginx and conform to some new standards

nemchik opened the pull request:
References:
https://mozilla.github.io/server-side-tls/ssl-config-generator/
https://observatory.mozilla.org/
https://securityheaders.io/
https://www.ssllabs.com/ssltest/

The former reference for the settings being used has been discontinued, so the above new references have been added and their recommended standards have been set.

nemchik commented:
FYI the former reference was https://cipherli.st/
The git repo for the former reference is https://github.com/RaymiiOrg/cipherli.st
There is a PR on the git repo to update the readme and explain that the site is gone https://github.com/RaymiiOrg/cipherli.st/pull/94

These new changes result in an A grade on https://securityheaders.io/ and a C+ grade on https://observatory.mozilla.org/
Getting a higher score on either site requires the end user to take some extra steps on their own.

TommyE123 commented:
Thanks for this update!

I'm thinking we should change the version number at the top of the file so its picked up when updating nginx and will get copied over. The only problem is the user will need to copy over the server_name line again.....

Unless we change the update to first get the server_name line, backup the conf file, copy over the new one, and then set the server_name back. That shouldn't be to hard to implement.

Hows that sound?!?

nemchik commented:
Definitely agree on the `server_name` line needing to be copied.

We currently don't have a `ssl_certificate` or `ssl_certificate_key` line in this config because those are added by Certbot, but it would be a very good idea to make sure they are copied over as well.

Here is an example of the bottom of my config after running certbot:
```
  ##
  # App Location Config
  ##

  include /etc/nginx/locations-enabled/*.atomic.conf;

    ssl_certificate /etc/letsencrypt/live/14th.nemchik.com/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/14th.nemchik.com/privkey.pem; # managed by Certbot
}
```
Those are the very last lines in my config, so really all Certbot does is add the two `ssl_` lines inside the `}`. Aside from that I really don't see much reason for anyone to modify the file with the exception of the three configs that have comments in the file about how to enable the extra security (Strict Transport, return 301, and dhparam) and those would need to be uncommented every time the file updates, but it's really not a huge issue (if they get re-commented it won't break anything, but if we lost the `ssl_` cert lines or the `server_name` that would break things).

nemchik commented:
Don't merge this. It breaks local access. I'll fix it and push a new commit and rebase soon.

nemchik commented:
I've rebased this and it's ready to merge (in my repo Travis is not failing on this commit after rebasing it).

We had discussed trying to preserve the `server_ name` line as well as any certificates placed by certbot, so that would be a good idea as well (not sure how to do that exactly).

nemchik committed:
Harden nginx and conform to some new standards

Ombi no longer requires the rewrite

Rename Atomic's nginx config, split into snippets, and use root directive only when needed

Nginx location version bumps and minor config fixes

Create nginx snippet folder if missing

Beautify all nginx config files

nemchik commented:
Per some other discussions we've had on gitter I've added some new things and changed this PR a bunch. This is ready to merge as of writing but we should still keep in mind that we probably need to account for the `server_name` line being changed by things like DuckDNS and the `ssl_certificate` or `ssl_certificate_key` lines being added by certbot. Those things would need to carry over when configs are updated.

Also this PR renamed the `AtoMiC-ToolKit-configured-sites` file to `server.atomic.conf` and moved it to a subfolder in the repo for the sake of consistency with the actual file structure after install. We should be able to safely remove the symlink from `/etc/nginx/sites-enabled/AtoMiC-ToolKit-configured-sites` and the file will remain in `/etc/nginx/sites-available/AtoMiC-ToolKit-configured-sites`
